### ISSUE TITLE
Pod-utils frequently flake

### ISSUE THREAD

**Issue by fejta** [kubernetes/test-infra#10884]
https://github.com/kubernetes/test-infra/pull/10882

```json
{
  "component":"sidecar",
  "error":"failed to upload to GCS: failed to upload to GCS: encountered errors during upload:
    [
      [Post https://www.googleapis.com/upload/storage/v1/b/kubernetes-jenkins/o?alt=json\u0026prettyPrint=false\u0026projection=full\u0026uploadType=multipart: oauth2: cannot fetch token: Post https://oauth2.googleapis.com/token: net/http: TLS handshake timeout]
      [Post https://www.googleapis.com/upload/storage/v1/b/kubernetes-jenkins/o?alt=json\u0026prettyPrint=false\u0026projection=full\u0026uploadType=multipart: oauth2: cannot fetch token: Post https://oauth2.googleapis.com/token: net/http: TLS handshake timeout]
      [Post https://www.googleapis.com/upload/storage/v1/b/kubernetes-jenkins/o?alt=json\u0026prettyPrint=false\u0026projection=full\u0026uploadType=multipart: oauth2: cannot fetch token: Post https://oauth2.googleapis.com/token: net/http: TLS handshake timeout]
    ]",
  "level":"error",
  "msg":"Failed to report job status",
  "time":"2019-01-22T22:05:04Z"
}
```
EDIT(cjwagner):
Failing to upload to GCS may be caused by CPU limitations, but it is worsened by a lack of retries which we should really have anyways so adding retries to the upload logic seems like the first step to addressing this.
This is the upload logic that should be retried: https://github.com/kubernetes/test-infra/blob/69cc5b2962487101421afee427c87eb1a8069114/prow/pod-utils/gcs/upload.go#L65-L72
Note that we'll need to create a new writer for each retry and that the group shouldn't be marked done until a retry has succeeded or all retries have been exhausted.

**Comment by fejta** [comment]
Officially a flake:

forever pending: https://gubernator.k8s.io/build/kubernetes-jenkins/pr-logs/pull/test-infra/10879/pull-test-infra-bazel/23533

pass: https://gubernator.k8s.io/build/kubernetes-jenkins/pr-logs/pull/test-infra/10879/pull-test-infra-bazel/23539

**Comment by stevekuznetsov** [comment]
@fejta and I also thought about this and it's most likely the uploader running out of CPU, so if we set CPU requests for the sidecar we can probably work around this

**Comment by jluhrsen** [comment]
@stevekuznetsov I submitted a PR to do the retries, although I did see your comment about the uploader
running out of CPU. I can try to address that as well (or instead of the PR just submitted), if you can
point me where that might be tweaked?

**Comment by stevekuznetsov** [comment]
To tweak the CPU issues, you'd need a Prow cluster to inspect and look at average use -- I have that data and this week I aim to fix that. Retry might not be reasonable for the CPU starvation case but would be useful in other cases, so let's do it regardless.

**Comment by stevekuznetsov** [comment]
`net/http: TLS handshake timeout` makes me think that you are running out of CPU on the container trying to do the upload, it can't calculate fast enough for the handshake. I would suggest using Prometheus to gather average cpu use for the initcontainer and add that as a request.

**Comment by stevekuznetsov** [comment]
Is this still happening?

**Comment by cjwagner** [comment]
Yes this is still happening and should be addressed with a retry loop and/or CPU limits on the sidecar container.
/open
/remove-lifecycle rotten
/help
/good-first-issue

**Comment by fejta** [comment]
If you want this to be a good first issue can you please update the description to include some links to the code that should be changed and how?

Right now the description is too ambiguous

**Comment by stevekuznetsov** [comment]
@alvaroaleman can you shed any light on this?

**Comment by alvaroaleman** [comment]
> @alvaroaleman can you shed any light on this?

No, I don't think I've seen this before. Adding requests and retrying sounds like a good idea.

**Comment by stevekuznetsov** [comment]
I thought we did add retries :|

**Comment by stevekuznetsov** [comment]
@cjwagner do you have a plot of this over time? I'm surprised I don't see it at all on my build clusters. I'm still suspicious of the message, too, which often indicates some sort of CPU throttle

**Comment by cjwagner** [comment]
> @cjwagner do you have a plot of this over time? I'm surprised I don't see it at all on my build clusters. I'm still suspicious of the message, too, which often indicates some sort of CPU throttle

I do not. Even if CPU throttling is the root cause of the issues we are seeing now, I think we should still add retries since we should really be retrying anyways and that will catch more problems than just CPU throttling. Networks are unreliable and the upload could fail for a lot of reasons.

What led us to believe that CPU throttling is the problem?

**Comment by fejta** [comment]
Network operations are guaranteed to occasionally fail at scale, and we should be robust to this reality.

**Comment by TAOXUY** [comment]
+1

**Comment by chizhg** [comment]
Is it possibly the same issue as https://github.com/knative/test-infra/issues/2081?

The error messages are slightly different, but they both have `oauth2: cannot fetch token: Post https://oauth2.googleapis.com/token`

**Comment by alvaroaleman** [comment]
@chizhg This issue here really is a simple "We have a lot of jobs that use podutils and sometimes things fail for random reasons so we should retry".

Crier has retrying already, so its a different story there.

**Comment by kushthedude** [comment]
Hey, @alvaroaleman Is this issue addressed? If not, I would like to look into it. As from discussions I can see that we need to introduce retry logic for podutils, right?

**Comment by alvaroaleman** [comment]
@kushthedude sorry, I completely missed your comment. This issue hasn't been addressed and yes, this is about adding retry logic in the podutils

**Comment by ajaygm** [comment]
@alvaroaleman - One more clarification needed...
The last commit on this topic was https://github.com/kubernetes/test-infra/issues/10884#ref-commit-b79bc0a (https://github.com/jluhrsen/test-infra/commit/b79bc0aa7cd10221d3bc609cf16bf111a62d7371)
Can you please have a look at that and see if that was sufficient, cause I didn't find any pull request with comments on that change.
If that looks sufficient, it could be a good first time contributor's attempt to learn the contribution process - I could try it :wink:.

**Comment by alvaroaleman** [comment]
Hi @ajaygm .

That commit looks good, the spontaneous three comments I'd have are:
* It adds a `lock` that serializes all uploads, we don't want that (We already added something to limit parallelization after that PR was created, but in order to limit memory usage)
* The `Info("Retrying upload")` log in the end will be shown even when it won't retry anymore when all retries failed
* Before we retry, we should introduce an exponential delay

**Comment by alvaroaleman** [comment]
Fixed in #19884
/close
